selftest/uprobe-multi/main.go

```diff
@@ -0,0 +1,238 @@
+		offsets = append(offsets, uint64(offset))
+		cookieToFunctionInfo[cookie] = FunctionInfo{
+			Name:   symbol.Name,
+			Offset: uint64(offset),
+		}
+	}
+
+	bpfModule, err := bpf.NewModuleFromFile(bpfProgramObject)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(-1)
+	}
+	defer bpfModule.Close()
+
+	if err = resizeMap(bpfModule, "events", 8192); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(-1)
+	}
+
+	log.Println("getting program")
+	prog, err := bpfModule.GetProgram(bpfProgramName)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(-1)
+	}
+
+	log.Println("setting expected attach type uprobe multi before loading")
+	if err = prog.SetExpectedAttachType(bpf.BPFAttachTypeTraceUprobeMulti); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(-1)
+	}
+
+	log.Println("loading object")
+	err = bpfModule.BPFLoadObject()
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(-1)
+	}
+
+	log.Println("attaching multi uprobes")
+	for i := 0; i < len(offsets); i += attachBatchSize {
+		end := i + attachBatchSize
+		if end > len(offsets) {
+			end = len(offsets)
+		}
+
+		_, err = prog.AttachUprobeMulti(-1, binaryPath, offsets[i:end], cookies[i:end])
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			os.Exit(-1)
+		}
+	}
+
+	log.Println("initializing events ring buffer")
+	eventsChannel := make(chan []byte)
+	rb, err := bpfModule.InitRingBuf("events", eventsChannel)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(-1)
+	}
+
+	rb.Poll(300)
+
+	// We get back from BPF and keep track of the function having traced via cookies.
+	log.Println("consuming events")
+	got := make(map[string]struct{})
+	go func() {
+		for {
+			b := <-eventsChannel
+			var event Event
+			buf := bytes.NewBuffer(b)
+			if err = binary.Read(buf, binary.LittleEndian, &event); err != nil {
+				// Error handling is out of scope for this test.
+				continue
+			}
+			cookie := event.Cookie
+			info, ok := cookieToFunctionInfo[cookie]
+			if !ok {
+				// Error handling is out of scope for this test.
+				continue
+			}
+			got[info.Name] = struct{}{}
+		}
+	}()
+	// Just wait for a minimum amount of time for the tested tracee to call
+	// the expected functions.
+	time.Sleep(2 * time.Second)
+
+	// Verify that all uprobes have been executed.
+	for _, symbolName := range expectedSymbolNames {
+		if _, ok := got[symbolName]; !ok {
+			fmt.Fprintf(os.Stderr, "function %s has not been traced\n", symbolName)
+			os.Exit(1)
+		}
+	}
+	log.Println("all functions have been traced")
+
+	// Test that it won't cause a panic or block if Stop or Close called multiple times
+	rb.Stop()
+	rb.Stop()
+	rb.Close()
+	rb.Close()
+	rb.Stop()
+}
+
+// getFunSyms returns the list of elf.Symbol of type function.
+func getFunSyms(name string) ([]elf.Symbol, error) {
+	var symbols []elf.Symbol
+
+	b, err := elf.Open(name)
+	if err != nil {
+		return nil, err
+	}
+	syms, err := b.Symbols()
+	if err != nil {
+		return nil, err
+	}
+	log.Println(syms)
```

Comment:
Instead of printing all:

```suggestion
	log.Printf("found %d symbols in %s\n", len(syms), name)
	log.Printf("showing first %d symbols\n", min(10, len(syms)))
	for i := 0; i < min(10, len(syms)); i++ {
		log.Printf("symbol %d: %v\n", i, syms[i])
	}
```